add support for mTLS in remote cache usage

tgolsson opened the pull request:
This PR adds support for mTLS when using remote caching. The API is fairly similar to what we've already have, but I've taken the liberty of doing some type cleanups. I also did some refactoring of the tls setup code to move some earliers to parse-time rather than conversion time, and simplify a bit.

I also found a bug with the `alpn` config, which isn't actually valid -- hyper just barfs when that is defined. Not sure how no-one else is hitting that, but it is what it is. Maybe a version/release issue. Comes from here: https://github.com/rustls/hyper-rustls/blob/163b3f539a497ae9c4fa65f55a8133234ef33eb3/src/connector/builder.rs#L46-L51.

I unfortunately don't have any non-mTLS cache to test against, so would love if someone else can test that I haven't broken something there.

(I need to figure out how to easily test-drive this in CI etc; right now I can only validate this works to read/write locally. Otherwise I guess I can test it for real in 2.19 :( )

tdyas commented on src/rust/engine/grpc_util/src/tls.rs:
```diff
@@ -8,18 +8,36 @@ use tokio_rustls::rustls::{Certificate, ClientConfig, Error, RootCertStore, Serv
-  pub fn new_without_mtls(root_ca_certs: Option<Vec<u8>>) -> Self {
-    Self {
+  /// Creates a new config with the given root CA certs and mTLS config.
+  pub fn new(root_ca_certs: Option<&[u8]>, mtls: MtlsConfig) -> Result<Self, String> {
```
Maybe just merge the two methods into a `new` which takes `root_ca_certs` and `mtls` both as options?

tdyas commented on src/rust/engine/grpc_util/src/tls.rs:
```diff
@@ -33,71 +51,108 @@ impl TryFrom<Config> for ClientConfig {
-    tls_config.alpn_protocols = vec![b"h2".to_vec()];
+    // Note[TSolberg]: This causes hyper to explode -- it seems a bit odd to pre-set negotiation,
+    // but this was added for a reason presumably.
+    //tls_config.alpn_protocols = vec![b"h2".to_vec()];
```
`tonic` sets it in its default setup logic presumably `hyper` does not. See https://github.com/hyperium/tonic/blob/6af8d5cc31a068d78c4a7937ad2eadb073e06025/examples/src/tls_rustls/server.rs#L44. I replicated that logic here since `tonic` had it.

If it is unnecessary now, then let's remove it but I would like to confirm that `hyper` does in fact set it before we do that and/or understand why `tonic` still sets it but it causes the error for you.

tgolsson replied:
We do set it here:

https://github.com/tgolsson/pants/blob/ts%2Fremote-cache-mtls/src/rust/engine/grpc_util/src/channel.rs#L51-L55
https://github.com/rustls/hyper-rustls/blob/163b3f539a497ae9c4fa65f55a8133234ef33eb3/src/connector/builder.rs#L221-L227

tgolsson replied:
But since the tlsconfig is used in many places I'm worried there's other paths that do not construct a channel in that way.

tdyas replied:
The code at that `hyper-rustls` link sets both `h2` and `http/1.1` as "next protocols" for TLS to try as part of ALPN. gRPC only runs over `h2` though so resetting `alpn_protocols` to `h2` may be the preferred action to take since gRPC can never run over http/1.1.

tgolsson replied:
After looking again; we *might* only be getting http2. This complex builder-pattern is a bit... hard to follow. I'll have to trace out what state transitions we'd actually be going through. This is the same function but lands with only http2.

https://github.com/rustls/hyper-rustls/blob/163b3f539a497ae9c4fa65f55a8133234ef33eb3/src/connector/builder.rs#L164C1-L175C6

tdyas replied:
Then seems like it should be safe to remove the `grpc_util` logic which sets that `alpn_protocols` then.

tgolsson replied:
Yeah. When I do a unit-test tomorrow I'll see if I can add a unit-test for that as well, after we've built the connector.

tdyas commented on src/rust/engine/process_execution/remote/src/remote.rs:
```diff
@@ -173,19 +175,25 @@ impl CommandRunner {
   ) -> Result<Self, String> {
-    let execution_use_tls = execution_address.starts_with("https://");
+    let needs_tls =
+      execution_address.starts_with("https://") || execution_address.starts_with("grpcs://");
```
A check for `grpcs://` is unnecessary here since the Python layer translates `grpcs` to `https`. See https://github.com/pantsbuild/pants/blob/eecd07908868af3e3e0d97a91852de2785b7b16c/src/python/pants/option/global_options.py#L467-L471.

tgolsson replied:
Hmm, I see your point, but it *looks* like a bug when we don't check it -- I know what I wrote in the config, and this doesn't check it. I'm wondering whether that rewrite should happen at the last possible moment, not far far away from where the value is processed?

tgolsson replied:
(This is a particular issue of stringly typed interfaces maybe. We don't care about the "protocol" spelling, we care about whether it's TLS. Maybe we should convert to a URI, then it'd be an enumeration so you can't write "wrong" code.)

tdyas replied:
Strongly typing this is fine by me.

huonw replied:
This protocol/address code is changing a lot with https://github.com/pantsbuild/pants/pull/19827, I've added tweaking this to https://github.com/pantsbuild/pants/issues/19902.

tgolsson replied:
@huonw Great, will revert and we'll tackle it separately to not scope-creep this PR.

tdyas commented on src/rust/engine/process_execution/remote/src/remote_cache/reapi.rs:
```diff
@@ -29,13 +29,25 @@ impl Provider {
-    let tls_client_config = if action_cache_address.starts_with("https://") {
-      Some(grpc_util::tls::Config::new_without_mtls(root_ca_certs).try_into()?)
+    let needs_tls =
+      action_cache_address.starts_with("https://") || action_cache_address.starts_with("grpcs://");
```
`grpcs://` check is unnecessary. See earlier comment about how `grpcs://` is translated to `https://`.

tdyas commented on src/rust/engine/process_executor/src/main.rs:
```diff
@@ -170,6 +170,16 @@ struct Opt {
   cas_root_ca_cert_file: Option<PathBuf>,
 
+  /// Path to file containing client certificates for the CAS server.
+  /// If not set, TLS will not be used when connecting to the CAS server.
```
Do you mean TLS won't be used or "TLS with a client certificate" won't be used?

tgolsson replied:
TLS -> mTLS.

tdyas commented on src/rust/engine/process_executor/src/main.rs:
```diff
@@ -170,6 +170,16 @@ struct Opt {
+  cas_mtls_certs_file: Option<PathBuf>,
+
+  /// Path to file containing client key for the CAS server.
+  /// If not set, TLS will not be used when connecting to the CAS server.
```
Similar comment about the wording as earlier.

tgolsson replied:
TLS -> mTLS.

tdyas commented on src/rust/engine/src/context.rs:
```diff
@@ -151,7 +153,9 @@ impl Core {
     remote_store_address: &Option<String>,
-    root_ca_certs: &Option<Vec<u8>>,
+    root_ca_certs: Option<&[u8]>,
+    mtls_certs: Option<&[u8]>,
+    mtls_key: Option<&[u8]>,
```
Maybe combine these into a single argument as an `Option<&[u8], &[u8]>` (or a struct if having names works better)? That way, the caller must ensure that both are specified (and can do error checks in the Python option parsing layer).

tdyas commented on src/rust/engine/src/context.rs:
```diff
@@ -513,6 +543,22 @@ impl Core {
+      .mtls_certs_path
+      .as_ref()
+      .map(|path| {
+        std::fs::read(&path).map_err(|err| format!("Error reading mtls certs file {path:?}: {err}"))
```
Nit: Maybe use `mTLS` for the naming for consistency instead of `mtls`?

tdyas commented on src/rust/engine/src/context.rs:
```diff
@@ -513,6 +543,22 @@ impl Core {
+      .mtls_key_path
+      .as_ref()
+      .map(|path| {
+        std::fs::read(&path).map_err(|err| format!("Error reading mtls key file {path:?}: {err}"))
```
Nit: Maybe use `mTLS` for the naming for consistency instead of `mtls`?

tdyas reviewed (commented):
Thank you for this. I left some comments.

Also, maybe also add a test in https://github.com/pantsbuild/pants/blob/eecd07908868af3e3e0d97a91852de2785b7b16c/src/rust/engine/grpc_util/src/channel.rs to ensure mTLS mode works?

tgolsson commented:
@tdyas Thanks for the comments! Updated from feedback and addressed some points in replies. Fixed some ergonomics to accept `AsRef` for the data as well in the configuration constructor, so it works as well with both owned and unowned data.

Re: tests -- as far as I can tell, those tests don't use anything at all from the code I've changed. I went ahead and added a test locally, but it doesn't pass. I think this happens because I can't configure "mtls but don't verify the certs" without re-implementing a custom rustls verifier, which seems... excessive. You added the certs; if you recall how they were generated I can create new ones and see if that solves the issue. :-)

tdyas commented:
> Re: tests -- as far as I can tell, those tests don't use anything at all from the code I've changed. I went ahead and added a test locally, but it doesn't pass. I think this happens because I can't configure "mtls but don't verify the certs" without re-implementing a custom rustls verifier, which seems... excessive. You added the certs; if you recall how they were generated I can create new ones and see if that solves the issue. :-)

I don't recall exactly. I may have copied the certs from `tonic` or `rustls` but they are actually technically expired and so invalid so the TLS integration test actually rejects them which is why I configured TLS in the test to not verify the certificates.

My request here is to ensure we have an integration test proving mTLS works since we have no such test and I would like to prevent future regressions here.

I did some searching tonight and found the [`rcgen` crate](https://crates.io/crates/rcgen) which appears to have the ability to easily generate self-signed certificates entirely in Rust code. I found it in this `rustls` issue: https://github.com/rustls/rustls/issues/400

tgolsson commented:
Will have a look! I'll do a test in `tls.rs` purely then, just validating that we set up the right config options. I think beyond that it's fairly out of our control, as we just pass the config to Hyper's connector builder.

tdyas commented:
> Will have a look, but I'll probably a test in tls.rs purely. Validating that we set up the right config options. I think beyond that it's fairly out of our control, as we just pass the config to Hyper's connector builder.

My thought is just for the integration test to be a "sanity check." So the verifier does not need to actually verify the certificates, just check that certificates were passed. My concern is we refactor the TLS setup logic in the future and somehow not even pass them to `rustls`.

tdyas commented:
fyi I added the label to cherry-pick this automatically to the `2.18.x` branch.

tgolsson commented:
Thank you! Will try to finish up tomorrow evening.

tgolsson commented:
@tdyas PTAL again! I added three tests: Does the cert reach the client, and does the MTLS generate an expected rustls::ClientConfig. I found no good way to test for alpn without breaking down the channel/client construction into ridiculously small steps. But we do both `enable_http2` and `http2_only(true)` so I think we're safe there.

Also reverted the `grpcs` check as discussed.

tdyas commented on docs/markdown/Using Pants/remote-caching-execution/remote-execution.md:
```diff
@@ -91,6 +91,9 @@ remote_execution_address = "grpcs://build.example.com:443"
 # This is optional, Pants will auto-discover certificates otherwise.
 remote_ca_certs_path = "/etc/ssl/certs/ca-certificates.crt"
+# this allows you to setup mTLS with a client certificate and key.
+remote_mtls_certs_path = "/etc/ssl/certs/client-cert.pem"
```
Naming questions:
1. Can there be multiple certificates in the file at that path or just one? I.e., should the option be `remote_mtls_cert_path` instead of `remote_mtls_certs_path`?
2. These options are really only about enabling the TLS client certificate and key. mTLS actually implies the superset of doing _mutual_ client cert verification and server cert verification. Should we rename these options to reflect that they only pertain to the client certificate? Bazel calls its equivalent options [`--tls_client_certificate`](https://bazel.build/reference/command-line-reference#flag--tls_client_certificate) and [`--tls_client_key`](https://bazel.build/reference/command-line-reference#flag--tls_client_key). Maybe we s should similarly call the Pants options `remote_tls_client_cert_path` and `remote_tls_client_key`?

tdyas replied:
cc @stuhood and @benjyw on input on the option naming

tgolsson replied:
1) Yes, I think it's always a chain formally.

2) Makes sense on naming, I had that during my initial setup but since the internal config calls it mtls and Pants is never a server I went for that name. But I'm happy to rename.

tdyas replied:
Probably worth asking for some opinions in Slack. I'll tag this comment thread there.

tdyas replied:
Also given the mTLS code was never visible to users, my view is we are free to name it as if this were a new feature. :)

benjyw replied:
Yeah, I think `client` in the name makes sense, presumably in an mTLS scenario the server cert still comes via the CA chain?

tgolsson replied:
Renamed all the public places that I introduced, and most validation messages. Kept the lower level structs as `mtls_data` to harmonize with `MtlsConfig`.

tgolsson replied:
Kept mTLS in the docs as a search needle, but also added `client authentication` so both are searchable.

tgolsson commented:
Just tested everything on our prod repo after the rebase, so should be G2G.
```
  remote_cache_requests: 9
  remote_cache_requests_cached: 9
```

benjyw commented on src/python/pants/option/global_options.py:
```diff
@@ -1602,6 +1608,35 @@ class BootstrapOptions:
+            """
+            Path to a PEM file containing client certificates used for verifying secure connections to
+            `[GLOBAL].remote_execution_address` and `[GLOBAL].remote_store_address` when using
+            client authentication.
```
```suggestion
            client authentication (mTLS).
```

benjyw commented on src/python/pants/option/global_options.py:
```diff
@@ -1602,6 +1608,35 @@ class BootstrapOptions:
+            """
+            Path to a PEM file containing a private key used for verifying secure connections to
+            `[GLOBAL].remote_execution_address` and `[GLOBAL].remote_store_address` when using
+            client authentication.
```
```suggestion
            client authentication (mTLS).
```

benjyw commented on src/rust/engine/fs/brfs/src/main.rs:
```diff
@@ -763,14 +763,22 @@ async fn main() {
   let runtime = task_executor::Executor::new();
 
+  let tls_config = match tls::Config::new(root_ca_certs, None) {
+    Ok(tls_config) => tls_config,
```
It would be more consistent with other code here to use `expect()`.

tgolsson replied:
Ten lines up is another error + exit, which is what I mirrored.

benjyw commented on src/rust/engine/grpc_util/src/tls.rs:
```diff
@@ -33,71 +49,105 @@ impl TryFrom<Config> for ClientConfig {
       CertificateCheck::DangerouslyDisabled => {
         let tls_config = tls_config.with_custom_certificate_verifier(Arc::new(NoVerifier));
-        if let Some(MtlsConfig { key, cert_chain }) = config.mtls {
+        if let Some(MtlsConfig { cert_chain, key }) = config.mtls {
```
Why change the order? It would be nice to have one consistent field order everywhere. If this makes more sense, why not change the order in the definition of MtlsConfig?

tgolsson replied:
Switched order.

benjyw commented on src/rust/engine/grpc_util/src/tls.rs:
```diff
@@ -33,71 +49,105 @@ impl TryFrom<Config> for ClientConfig {
+          key = Some(rustls::PrivateKey(buf))
+        }
+        Item::X509Certificate(_) => unreachable!("filtered above"),
+        _ => todo!("non-exhaustive match"),
```
Should probably be a more useful end-user message?

tgolsson replied:
There's no other options, this is because the rustls_pemfile item enum is non-exhaustive. I'll change it to unreachable. Unfortunately since the items are `Item::Foo(Vec<u8>)` printing *what* was found would dump raw bytes to the console.

benjyw commented on src/rust/engine/fs/fs_util/src/main.rs:
```diff
@@ -369,7 +369,7 @@ async fn execute(top_match: &clap::ArgMatches) -> Result<(), ExitError> {
               )
             })?;
-            Some(MtlsConfig { key, cert_chain })
+            Some((cert_chain, key))
```
A pair of items of the same type, where order matters, makes me uneasy. Especially since we're not consistent with the order (see below). Could this be a struct with named fields, say `cert_chain_pem` and `key_pem`?

tdyas replied:
I agree. We should put this back as a `struct` with two named fields.

tgolsson replied:
There's inconsistent feedback; I had them as two fields and was asked to fold them into one option, now it's the other way around. I also did explicitly construct an MtlsConfig from the two pieces, and passed it to TlsConfig. I think "config objects" are quite weird when this local, and what *should* be done is to construct the TlsConfig immediately after loading the data and not in many places.

I believe this is best tackled as part of point 3 here: https://github.com/pantsbuild/pants/issues/19902.

tdyas commented on src/rust/engine/grpc_util/src/tls.rs:
```diff
@@ -33,71 +49,105 @@ impl TryFrom<Config> for ClientConfig {
+      }
+    }
+
+    let key = key.ok_or_else(|| "No private key found in mTLS key file".to_owned())?;
```
Probably want to replace `mTLS` with the name of the option for the key file.

tgolsson replied:
This place does not know that because there's many `main.rs` that reach this point. With proper setup of error chaining (formatting, expect, etc) it should be handled by higher levels.

tdyas commented on src/python/pants/option/global_options.py:
```diff
@@ -1602,6 +1608,35 @@ class BootstrapOptions:
             """
         ),
     )
+    remote_client_certs_path = StrOption(
```
Any reason for not including "tls" in the name similar to how Bazel names it? I.e., `remote_tls_client_certs_path` and `remote_tls_client_key_path`? I guess the choice was made with `remote_ca_certs_path` not including `tls` in the name.

tgolsson replied:
Yeah; I think all fields should have `tls` or none, yes.

tgolsson commented:
@benjyw @tdyas Responded to feedback again, PTAL. I'm happy to keep iterating but I think there's some overlap between what I'm doing and what @huonw is looking at with simplifying the remote. If we want to simplify TLS setup I do think the best approach is to construct it immediately after reading files, and pass `Option<TlsConfig>` to any interior APIs -- I'm not fundamentally doing anything other than what was *already* being done with `root_ca_certs`, and I think any arguments for or against my approach apply there as well.

That seems better to tackle holistically, and not apply band-aids. I'm happy to tackle "construct TLSConfig once" in a separate PR if you want, but that mostly depends on what @huonw wants to achieve.

tdyas commented:
> That seems better to tackle holistically, and not apply band-aids. I'm happy to tackle "construct TLSConfig once" in a separate PR if you want, but that mostly depends on what @huonw wants to achieve.

I'm fine with the internal redesign proceeding in other PRs. I'm just concerned about the switch from `struct MtlsConfig` to a 2-tuple but if that will be refactored away than I'm less concerns.

@huonw: What are your thoughts?

huonw commented:
Yeah, more than happy to have whatever internal redesign happen after landing, to get the user-facing aspects in earlier. Feel free to adjust #19902 to include extra things too.

benjyw commented:
@tgolsson The auto-cherry-pick to 2.18.x failed. Can you do that manually? The steps are listed above.

tgolsson commented:
Ack, will look tonight!
